### ISSUE TITLE
Pathfinder Variational Inference 

### ISSUE THREAD

**Issue by fmkroci** [NIFTy-PPL/NIFTy#40]
Hello,

I have quite a large model with the additional complication that I don't have a simple way to calculate the FIM of the likelihood. 
So far I have simply used a local quadratic approximation of the likelihood to find a rough approximation for the FIM. 
I can show in simulated data that this works well, but calculating the Hessian is quite expensive for the large amount of data that I have.

Therefore, I was looking for alternatives. I tried to understand if using Quasi-Newton methods can give you a way of making use of the likelihood/gradient evaluations during KL optimization to roughly estimate the FIM.
I stumbled upon this paper: 
"Pathfinder: Parallel quasi-Newton variational inference"
https://arxiv.org/abs/2108.03782
Which, I believe, is what I was looking for.

From my understanding, blackjax has an implementation of the Pathfinder algorithm.
So the approach should, in theory, be available for nifty.re.

To finally, arrive at my question(s).
Do you think this is worthwhile to try with nifty? (Or has anyone already tried?)

If you think it makes sense to test it, is there an example of how to run blackjax with nifty.re.
So far, my code is using the classic nifty and due to limited time I could not make the switch no nifty.re yet.

I hope this approach makes sense in the first place :)

**Comment by ph-frank** [comment]
Thank you for your question and interest in nifty(.re)!
Indeed, as you point out very nicely, depending on the problem setup a different sampler/optimizer then the ones provided by Nifty natively may be more suitable to your needs. Pathfinder is widely used and can be a fast and powerful alternative in your problem setup. Note that it provides a Gaussian approximation to the posterior however, whereas some VI variants in NIFTy are non-Gaussian. 

Fortunately interfacing with other jax packages is extremely simple and accessible, as all objects in nifty.re remain jax transformable (as is the case for most jax based packages).
Specifically for interfacing with blackjax, the object of interest in the energy (negative log probability of parameters and data), commonly referred to as `logdensity_fn` in blackjax. A jax transformable `logdensity_fn` is the main ingredient required to set up the sampler. Nifty provides fully jax transformable likelihood energies, and parameters are re-parameterized to be standard normal distributed a priori, providing a fully jax transformable energy by adding likelihood and prior (see [re.optimize_kl._StandardHamiltonian](https://github.com/NIFTy-PPL/NIFTy/blob/2fa69f8613b430287e0cf27b4893349ae74ab93d/src/re/optimize_kl.py#L64).

Also, if you want to give it a go, feel free to report on your experiences in interfacing! I am also happy to have a look at any PR that demonstrates it in more detail :)

**Comment by fmkroci** [comment]
Thanks a lot for the detailed answer and the quick response!

So, if MGVI works well, there is a good chance that Pathfinder might also work well, correct?

I will try to get a simple demo running using nifty.re, when I find the time.

Since my analysis is built around classic/legacy NIFTy, is there an existing wrapper (maybe using jaxbind) to use an existing Hamiltonian with the optimizer implementations of nifty.re?

**Comment by ph-frank** [comment]
I fear I am not familiar enough with the details of your problem and the samples to judge that, they both provide a non-diagonal Gaussian approximation although the way the covariance is being build is quite different.
We do not provide a direct wrapper, but if you want to expose pure numpy NIFTy code to jax, `jaxbind` should provide the necessary functionality.

**Comment by fmkroci** [comment]
No worries thanks a lot for the help!
I will check how easily I can wrap the existing code. 
Once I have a simple example I can make a PR, if you are interested.

**Comment by roth-jakob** [comment]
If you could open a PR for a generic wrapper of numpy NIFTy operators into nifty.re that would be awesome! I think that should be quite straightforward. Feel free to tag me if you run into any problems.

Note: Likelihoods are handled somewhat differently in numpy NIFTy and NIFTy.re. I would propose to have a wrapper only for the normal operators but not likelihoods.  Thus, in your cod,e you could wrap your entire forward model and would just need to instantiate a nifty.re likelihood to which you amend your model.
